### DISCUSSION TITLE
`mkdocs` vs `sphinx`

### DISCUSSION THREAD

**Discussion by davidorme** [ImperialCollegeLondon/virtual_rainforest#5] [Q&A] [answered]
## Background

This is a really handwavey discussion topic... Essentially, I have opinions that don't have a neat resolution:

### I dislike ReStructuredText (RST) as a markup language and like Markdown. 

Clearly the classic Markdown (https://daringfireball.net/projects/markdown/) is much more limited that RST but more complex extensions of Markdown abound. Specifically MyST (https://myst-parser.readthedocs.io/en/latest/) is specifically designed to be an extended Markdown that forms a Sphinx-compatible replacement.

### Literate programming in documentation

We are going to need to want **how-to guides** in our documentation and these should be [literate programming](https://en.wikipedia.org/wiki/Literate_programming) documents. With Python, the natural choice is Jupyter Notebooks (https://jupyter.org/), which can sit happily within our `docs` and get run during documentation builds. 

This handily also checks that example code in documentation _actually_ runs.

### I dislike the `.ipynb` format for Jupyter Notebooks

The native format for Jupyter Notebooks is `.ipynb`. This is a JSON formatted file - it is _technically_ human readable and usually you would work with a rendered version in a Notebook viewer or HTML - but it isn't really directly editable.

A solution exists, which is that MyST (and many other formats) can be used in Jupyter Notebook via [Jupytext](https://jupytext.readthedocs.io/en/latest/install.html) and can be built in Sphinx via [Myst-NB](https://myst-nb.readthedocs.io/en/latest/)

### But... the MyST project currently still requires RST in docstrings for autodoc

MyST sits on top of `sphinx` for documentation generation and the really strong `autodoc` features of `sphinx` are hard coded to expect docstrings in RST. Mixing markup languages between docstrings and other documentation is best avoided. 

There are two relevant things here:

1.  The `MyST` developers are well aware that this isn't great and there is a lot of active development:

    https://github.com/executablebooks/MyST-Parser/issues/228

2. There is also `mkdocs`, which is an alternative documentation generator to `sphinx`. It is lighter-weight and really promising but isn't quite as mature. It does have an existing Markdown based autodoc extension, but the whole project isn't quite as mature as `sphinx`.

## Questions

Vague and poorly defined. Thoughts?

**Comment by dalonsoa** [comment, marked as answer]
Given that Sphinx is very mature and has a lot of capabilities, and considering that the MyST people is working on the issue, I'd go for Sphinx and trust that, at some point, this will be sorted. 

Along the same lines, I'd suggest to use Google style docstrings and use the [Napoleon extension](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) for Sphinx to parse them properly in the API documentation.

**Reply by davidorme** [reply]
Completely agree on Google Style docstrings and Napoleon - I'm already using them in another project. Only wish I could use Myst docstrings 😄
